go/vt/vttablet/tabletserver/vstreamer/helper_event_test.go

```diff
@@ -0,0 +1,569 @@
+	for _, v := range values {
+		v2 := strings.Trim(v, "'")
+		for i, m := range meta {
+			m2 := strings.Trim(m, "'")
+			if m2 == v2 {
+				switch col.dataTypeLowered {
+				case "set":
+					bits |= 1 << uint(i)
+				case "enum":
+					bits = int64(i) + 1
+				}
+			}
+		}
+	}
+	return strconv.FormatInt(bits, 10)
+}
+
+func (ts *TestSpec) getRowEvent(table string, bv map[string]string, fe *TestFieldEvent, stmt sqlparser.Statement, flags uint32) string {
+	ev := &binlogdata.RowEvent{
+		TableName: table,
+		RowChanges: []*binlogdata.RowChange{
+			{
+				Before: nil,
+				After:  nil,
+			},
+		},
+		Flags: flags,
+	}
+	var row query.Row
+	for i, col := range fe.cols {
+		if fe.cols[i].skip {
+			continue
+		}
+		if col.dataTypeLowered == "binary" {
+			bv[col.name] = strings.TrimSuffix(bv[col.name], "\\0")
+		}
+		val := []byte(bv[col.name])
+		l := int64(len(val))
+		if col.dataTypeLowered == "binary" {
+			for l < col.len {
+				val = append(val, "\x00"...)
+				l++
+			}
+		}
+		row.Values = append(row.Values, val...)
+		row.Lengths = append(row.Lengths, l)
+	}
+	ev.RowChanges = ts.getRowChanges(table, stmt, &row)
+	vEvent := &binlogdata.VEvent{
+		Type:     binlogdata.VEventType_ROW,
+		RowEvent: ev,
+	}
+	return vEvent.String()
+}
+
+func (ts *TestSpec) getRowChanges(table string, stmt sqlparser.Statement, row *query.Row) []*binlogdata.RowChange {
+	var rowChanges []*binlogdata.RowChange
+	var rowChange binlogdata.RowChange
+	switch stmt.(type) {
+	case *sqlparser.Insert:
+		rowChange.After = row
+		ts.setCurrentState(table, row)
+	case *sqlparser.Update:
+		rowChange = *ts.getRowChangeForUpdate(table, row)
+		ts.setCurrentState(table, row)
+	case *sqlparser.Delete:
+		rowChange.Before = row
+		ts.setCurrentState(table, nil)
+	}
+	rowChanges = append(rowChanges, &rowChange)
+	return rowChanges
+}
+
+func (ts *TestSpec) getRowChangeForUpdate(table string, newState *query.Row) *binlogdata.RowChange {
+	var rowChange binlogdata.RowChange
+	var bitmap byte
+	var before, after query.Row
+
+	currentState := ts.getCurrentState(table)
+	if currentState == nil {
+		return nil
+	}
+	var currentValueIndex int64
+	var hasSkip bool
+	for i, l := range currentState.Lengths {
+		skip := false
+		isPKColumn := false
+		for _, pkColumn := range ts.pkColumns[table] {
+			if pkColumn == ts.fieldEvents[table].cols[i].name {
+				isPKColumn = true
+				break
+			}
+		}
+		if ts.options.noblob {
+			switch ts.fieldEvents[table].cols[i].dataTypeLowered {
+			case "blob", "text":
+				currentValue := currentState.Values[currentValueIndex : currentValueIndex+l]
+				newValue := newState.Values[currentValueIndex : currentValueIndex+l]
+				if string(currentValue) == string(newValue) {
+					skip = true
+					hasSkip = true
+				}
+			}
+		}
+		if skip && !isPKColumn {
+			before.Lengths = append(before.Lengths, -1)
+		} else {
+			before.Values = append(before.Values, currentState.Values[currentValueIndex:currentValueIndex+l]...)
```

Comment:
Do we know that `currentValueIndex+l` won't ever be out of bounds? I don't think it should be, but might be worth a check.

Reply:
Yes, the related lengths are in the proto.